controller/array_action/array_mediator_ds8k.py

```diff
@@ -370,17 +395,136 @@ def unmap_volume(self, volume_id, host_name):
+        logger.info(
+            "creating target api volume '{0}' from source volume '{1}'".format(target_volume_name,
+                                                                               source_volume_name))
+        source_api_volume = self._get_api_volume_by_name(source_volume_name)
+        capabilities = {config.CAPABILITIES_SPACEEFFICIENCY: None}
+        size_in_bytes = int(source_api_volume.cap)
+        pool = source_api_volume.pool
+        return self.create_volume(target_volume_name, size_in_bytes, capabilities, pool)
+
+    def _create_flashcopy(self, source_volume, target_volume):
+        logger.info(
+            "creating FlashCopy Mapping from '{0}' to '{1}'".format(source_volume.volume_name,
+                                                                    target_volume.volume_name))
+        source_volume_id = guid_to_ds8k_request_id(source_volume.id)
+        target_volume_id = guid_to_ds8k_request_id(target_volume.id)
+        volume_pair = flashcopy_request_volume_pair_parser(source_volume_id, target_volume_id)
+        try:
+            api_flashcopy = self.client.create_flashcopy(volume_pairs=volume_pair,
+                                                         options=["persistent"])
+        except (exceptions.ClientError, exceptions.ClientException) as ex:
+            if ERROR_CODE_ALREADY_FLASHCOPY in str(ex.message).upper():
+                raise array_errors.SnapshotAlreadyExists
+            elif ERROR_CODE_VOLUME_NOT_FOUND_OR_ALREADY_PART_OF_CS_RELATIONSHIP in str(
+                    ex.message).upper():
+                raise array_errors.VolumeNotFoundError
+            else:
+                raise array_errors.FlashcopyCreationError('{}:{}'.format(source_volume_id,
+                                                                         target_volume_id))
+        if not self.validate_flashcopy(api_flashcopy.id):
+            # #TODO Delete flashcopy CSI-1340
+            raise array_errors.FlashcopyCreationError(api_flashcopy.id)
+        return self._get_api_volume_by_id(target_volume_id)
+
+    def _delete_target_volume_if_exist(self, target_volume_name):
+        target_api_volume = self._get_api_volume_by_name_if_exists(target_volume_name)
+        if target_api_volume:
+            self._delete_volume(target_volume_name, not_exist_err=False)
+
+    def _create_snapshot(self, target_volume_name, source_volume_name):
+        target_volume = self._create_similar_volume(target_volume_name, source_volume_name)
+        source_volume = self._get_volume_by_name(source_volume_name)
+        try:
+            return self._create_flashcopy(source_volume, target_volume)
+        except (array_errors.VolumeNotFoundError, array_errors.SnapshotAlreadyExists) as ex:
+            logger.error("Failed to create snapshot '{0}': {1}".format(target_volume_name, ex))
+            self._delete_target_volume_if_exist(target_volume_name)
```

Comment:
Eran said he wanted to do rollback with retries.
You can see examples array_mediator_interface.py

Reply:
can you specify what method should be on retry?

Reply:
_delete_target_volume_if_exist

Reply:
done